Transform f1_elctrc_oper_rev

zschira opened the pull request:
<!--

Making a PUDL Pull Request

Before making a PR you may want to check out our:

* contributing guidelines: https://catalystcoop-pudl.readthedocs.io/en/latest/CONTRIBUTING.html
* code of conduct: https://catalystcoop-pudl.readthedocs.io/en/latest/code_of_conduct.html
* development process: https://catalystcoop-pudl.readthedocs.io/en/latest/dev/index.html

## PR Process Overview

* PRs have to get an approving review before merging into their development branch.
* Most PRs should be made against the `dev` branch, unless they are part of some larger ongoing refactoring, in which case there will be a persistent development branch for that work.
* It is much easier to do timely code reviews on smaller chunks of code. We try to keep PRs under 500 lines of code.
* Draft PRs are a good way to get early feedback on designs or several incremental commits that will add up to larger changes. If you want a review of a draft PR, make sure you contact the reviewer directly or mention their username in the PR comment, so they get a notification.
* How quickly we can review a PR will depend on how large and complex it is, and how busy we are, but ideally we strive to get an initial review done within a week. If there are going to be delays, we should at least comment on the PR to let you know the situation.
* If you believe you've addressed a reviewer's comments, respond with a brief note and mark the comment resolved. If further discussion is requried respond and do not resolve the comment.
* Before a PR is merged all reviewer comments should be resolved. If a reviewer doesn't feel that their comment has been sufficiently addressed, they may unresolve a comment.
* Be careful not to accidentally "start a review" when responding to comments! If this does happen, don't forget to submit the review you've started so the other PR participatns can see your comments (they are invisible to others if marked "Pending").
* In the period after an initial review when there is significant back-and-forth with the reviewer deciding what changes should actually be made, there should probably be daily interaction. If significant changes are required, it's usually best to request another review after those changes have been made.

Feel free to delete the commented-out parts of the template before submitting the PR.

-->

# PR Overview
This PR covers the transformation of the `f1_elctrc_oper_rev` table. Most of the work for the structured portion of this table was just reshaping and mapping to `electric_operating_revenues_300_duration`. There's no instant table in the XBRL data, so no merging is necessary. Row #'s 22-25 in the DBF table contain unstructured data that is siloed in the XBRL data to the `electric_operating_revenues_other_300_duration`. This unstructured data is not handled by this PR.
<!--

Include a short narrative summary of what's going on in the PR. This can be a bulleted list. You might want to include:

* What are you changing and why?
* Are there any known unsolved problems remaining in the PR?
* Is there anything that you want a reivewer to pay particular attention to?
* What kind of feedback are you looking for on the PR?

-->

# PR Checklist

- [x] Merge the most recent version of the branch you are merging into (probably `dev`).
- [x] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [x] Make sure you've included good docstrings.
- [x] Include unit tests for new functions and classes.
- [x] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

zschira commented on src/pudl/transform/params/ferc1.py:
```diff
@@ -3448,6 +3448,119 @@
+                    col: f"{col}_electric_operating_revenue"
+                    for col in [
+                        "residential_sales",
+                        # Note: the small_or_commercial/large_or_industrial columns are already prefixed
```
Not sure why, but these two columns have the prefix `electric_operating_revenue` already, which is a bit inconsistent with the rest of the columns

cmgosnell replied:
lol. that is weird but also is consistent with FERCs inconsistency.

cmgosnell commented on src/pudl/metadata/fields.py:
```diff
@@ -963,6 +967,11 @@
         "type": "boolean",
         "description": "Can the generator burn multiple fuels?",
     },
+    "mwh_sold": {
+        "type": "number",
+        "description": "Megawatt hours sold.",
+        "unit": "MWh",
+    },
```
I'd suggest using `sales_mwh` instead here which is already defined. I think this is currently only being used for EIA, but it seems analogous.

zschira replied:
Changed

cmgosnell commented on src/pudl/metadata/resources/ferc1.py:
```diff
@@ -703,6 +703,28 @@
+                "report_year",
+                "record_id",
+                "revenue_type",
+                "amount",
```
if this $'s are always revenues, could this be `revenue` instead? We really should settle on how we name all of these $ columns re: #2179 .. bc we haven't really settled on anything I do think `amount` is fine if you prefer it!

zschira replied:
I actually prefer `revenue`, and had considered that initially, but went with `amount` to try and be consistent with other tables. I've changed to `revenue` now.

cmgosnell reviewed (changes requested):
It seems like the most weird part of this was the duration rename. Its very encouraging that the main work here looks like it was adding params and the dbf/xbrl row maps!

So this table has a top chunk of electricity sold w/ revenue plus MWh sold and # of customers and then a second chunk with just revenues? I think it would be nice to put that content into the resource description. So users (and us!) knows why a chunk of the MWhs & # of customer fields are null.

Or are the ones that are being dropped in the `wide_to_tidy` all of these columns in the "Other Operating Revenue" section? In these tables that have structured and unstructured portions coming from DBF we've typically pulled all of the columns that are in the XBRL structured table even if they seem like they should be the total value of the unstructured bit of the table.

zschira commented on src/pudl/package_data/settings/etl_fast.yml:
```diff
@@ -73,8 +74,12 @@ ferc_to_sqlite_settings:
       - retained_earnings_118_duration
       - retained_earnings_118_instant
-      - statement_of_cash_flows_120_duration
+      - electric_operating_revenues_300_duration
+      - electric_operating_revenues_300_instant
+      - electric_operating_revenues_other_300_duration
+      - electric_operating_revenues_300_instant
```
I was using the `other` tables (and had a typo) when I was messing around with the unstructured data, but don't need them now, so I removed these

zschira commented on src/pudl/transform/params/ferc1.py:
```diff
@@ -3448,6 +3448,119 @@
+                    "mwh_sold",
+                    "avg_customers_per_month",
+                ],
+                "expected_drop_cols": 14,
```
These were actually columns that didn't have the mwh sold and average customers, that I was leaving out while deciding how to handle them, and forgot to add them back in. I've added these columns back, and had to add a small bit of code to drop some records with duplicate primary keys. The records in question were all from one utility in 2011 and every one of them had a revenue of 3.33e8 or 3.333e9 and all other fields were Null. These seemed pretty meaningless, so I dropped them all.

zschira commented:
> So this table has a top chunk of electricity sold w/ revenue plus MWh sold and # of customers and then a second chunk with just revenues? I think it would be nice to put that content into the resource description. So users (and us!) knows why a chunk of the MWhs & # of customer fields are null

Yeah that's correct. I've added a note in the description about the NULL fields on many records, and added notes in `dbf_to_xbrl.csv` indicating which `revenue_type`'s this applies to. Does this seem sufficient?

cmgosnell reviewed (approved):
this looks great! glad you slurped the revenue-only bit back in and I think your description in the resources is sufficient for sure!! approved (pending you merge dev in and the tests pass of course!)
